libgraphics/src/context.cpp

```diff
@@ -1350,34 +1463,42 @@ void MCGContextEnd(MCGContextRef self)
 	t_child_layer = self -> layer;
 	self -> layer = self -> layer -> parent;
 	
-	const SkBitmap& t_child_bitmap = t_child_layer -> canvas -> getTopDevice() -> accessBitmap(false);
-	
 	SkPaint t_paint;
 	t_paint . setAlpha((U8CPU)(self -> state -> opacity * 255));
 	
-	SkXfermode *t_blend_mode;
-	t_blend_mode = MCGBlendModeToSkXfermode(self -> state -> blend_mode);
-	if (t_blend_mode != NULL)
-	{
-		t_paint . setXfermode(t_blend_mode);
-		t_blend_mode -> unref();
-	}
+	t_paint.setBlendMode(MCGBlendModeToSkBlendMode(self->state->blend_mode));
 	
 	if (t_child_layer -> has_effects)
 	{
 		bool t_in_layer;
 		t_in_layer = false;	
 		if (self -> state -> opacity != 1.0f || self -> state -> blend_mode != kMCGBlendModeSourceOver)
 		{
-			self -> layer -> canvas -> saveLayer(NULL, &t_paint, (SkCanvas::SaveFlags) (SkCanvas::kHasAlphaLayer_SaveFlag | SkCanvas::kFullColorLayer_SaveFlag));
+			self -> layer -> canvas -> saveLayer(NULL, &t_paint);
 			t_in_layer = true;
 		}
 		MCGContextRenderEffects(self, t_child_layer, t_child_layer -> effects);
 		if (t_in_layer)
 			self -> layer -> canvas -> restore();
 	}
+	else if (t_child_layer->m_surface == nil)
+	{
+		// Software-backed layer
+		const SkBitmap& t_child_bitmap = t_child_layer->canvas->getTopDevice()->accessBitmap(false);
+		self->layer->canvas->save();
+		self->layer->canvas->resetMatrix();
+		self->layer->canvas->drawBitmap(t_child_bitmap, t_child_layer->origin_x, t_child_layer->origin_y, &t_paint);
+		self->layer->canvas->restore();
+	}
 	else
-		self -> layer -> canvas -> drawSprite(t_child_bitmap, t_child_layer -> origin_x, t_child_layer -> origin_y, &t_paint);
+	{
+		// Hardware-backed layer - copy it into RAM
+		sk_sp<SkImage> t_image = t_child_layer->m_surface->makeImageSnapshot();
+		self->layer->canvas->save();
+		self->layer->canvas->resetMatrix();
+		self->layer->canvas->drawImage(t_image, t_child_layer->origin_x, t_child_layer->origin_y, &t_paint);
```

Comment:
Unfortunately, the resetMatrix bit is required - I did try it that way before and it caused issues.